W3ChampionsStatisticService/Services/Tracing/Interceptors/TracingInterceptor.cs

```diff
@@ -0,0 +1,98 @@
+using Castle.DynamicProxy;
+using W3C.Domain.Tracing;
+namespace W3ChampionsStatisticService.Services.Interceptors;
+
+public class TracingInterceptor(ActivitySource activitySource) : IInterceptor
+{
+    private readonly ActivitySource _activitySource = activitySource;
+
+    public void Intercept(IInvocation invocation)
+    {
+        // 1. Check for [NoTrace] on the method first
+        var noTraceAttributeOnMethod = invocation.MethodInvocationTarget.GetCustomAttribute<NoTraceAttribute>()
+                                     ?? invocation.Method.GetCustomAttribute<NoTraceAttribute>();
+
+        if (noTraceAttributeOnMethod != null)
+        {
+            invocation.Proceed(); // [NoTrace] found, proceed without tracing.
+            return;
+        }
+
+        // 2. Check for [Trace] on the method
+        var methodTraceAttribute = invocation.MethodInvocationTarget.GetCustomAttribute<TraceAttribute>()
+                                 ?? invocation.Method.GetCustomAttribute<TraceAttribute>();
+
+        TraceAttribute effectiveTraceAttribute = methodTraceAttribute; // Prioritize method attribute
+
+        // 3. If no [Trace] on method, check for [Trace] on the class (TargetType)
+        if (effectiveTraceAttribute == null)
+        {
+            effectiveTraceAttribute = invocation.TargetType.GetCustomAttribute<TraceAttribute>();
+        }
+
+        if (effectiveTraceAttribute == null)
+        {
+            invocation.Proceed(); // No [Trace] attribute on method or class, proceed without tracing.
+            return;
+        }
+
+        string className = invocation.TargetType.Name;
+        string methodName;
+
+        if (methodTraceAttribute != null)
+        {
+            // [Trace] is on the method, OperationName is correctly set by [CallerMemberName]
+            methodName = methodTraceAttribute.OperationName;
+        }
+        else
+        {
+            // [Trace] is on the class, so use the actual invoked method's name for the span's method part.
+            // effectiveTraceAttribute.OperationName (from class) is ignored here for the method part of the span name.
+            methodName = invocation.Method.Name;
+        }
+
+        string fullSpanName = $"{className}.{methodName}";
+
+        using var activity = _activitySource.StartActivity(fullSpanName, ActivityKind.Internal);
+
+        try
+        {
+            if (activity != null)
+            {
+                ParameterInfo[] parameters = invocation.Method.GetParameters();
+                for (int i = 0; i < invocation.Arguments.Length; i++)
+                {
+                    // Check if the parameter has [NoTrace]
+                    if (parameters[i].GetCustomAttribute<NoTraceAttribute>() == null)
+                    {
+                        var arg = invocation.Arguments[i];
+                        var paramName = parameters[i].Name;
+                        activity.SetTag($"param.{paramName}", arg?.ToString() ?? "null");
```

Comment:
How does this handle function arguments?

Reply:
The way it works is it iterates through all function arguments and adds them to the span as "param.searchQuery", for example. The value is always the result of ToString() - unless the NoTrace annotation is set.

Reply:
As clarified via PM, the question was how functions as parameters (to allow dynamic invocations) are handled. The answer is that they aren't, they'll just be converted to string and have no meaningful value or instrumentation.

Can add that if the need arises.